apps/email/js/html_cache_restore.js

```diff
@@ -102,103 +250,236 @@ window.htmlCacheRestorePendingMessage = [];
+    if (startingUp) {
+      startupData.view = 'compose';
+      hydrateHtml(startupData.view);
     }
-    if (navigator.mozHasPendingMessage('notification')) {
-      window.htmlCacheRestorePendingMessage.push('notification');
+
+    dispatch('activity', [req]);
+  }
+
+  function onNotification(msg) {
+    console.log('mozSetMessageHandler: received a notification');
+    // Skip notification events that are not from a notification "click". The
+    // system app will also notify this method of any close events for
+    // notifications, which are not at all interesting.
+    if (!msg.clicked) {
+      if (startingUp) {
+        window.close();
+      }
+      return;
+    }
+
+    // Bail early if notification is ignored. Do this before the notification
+    // close() work, so that user has the opportunity to tap on the notification
+    // later and still activate that notification flow.
+    if (messageTooFast('notification') || ignoreMessage('notification')) {
+      return;
+    }
+
+    // Need to manually get all notifications and close the one that triggered
+    // this event due to fallout from 890440 and 966481.
+    if (typeof Notification !== 'undefined' && Notification.get) {
+      Notification.get().then(function(notifications) {
+        if (notifications) {
+          notifications.some(function(notification) {
+            // Compare tags, as the tag is based on the account ID and
+            // we only have one notification per account. Plus, there
+            // is no "id" field on the notification.
+            if (notification.tag === msg.tag && notification.close) {
+              notification.close();
+              return true;
+            }
+          });
+        }
+      });
     }
+
+    // Adjust the startupData view as desired by the notification. For upgrade
+    // cases where a previous notification from an older version of email
+    // used the iconUrl, this just means we will got to the message_list instead
+    // of the message_reader for the single email notification case, but that is
+    // OK since it is a temporary upgrade issue, and the email will still be
+    // seen since it should be top of the list in the message_list.
+    var view = msg.data.type;
+    if (startingUp && view) {
+      startupData.view = view;
+      hydrateHtml(view);
+    }
+
+    // The notification infrastructure does not automatically bring the app to
+    // the foreground, so if still hidden, show it now. Ideally this would not
+    // use a setTimeout, but it was getting incorrect document.hidden values on
+    // startup, where just a bit later the value does seem to be set correctly.
+    // The other option was to do this work inside the notification handler in
+    // mail_app, but the delay is long enough waiting for that point that the
+    // user might be concerned they did not correctly tap the notification.
+    setTimeout(function() {
+      if (document.hidden && navigator.mozApps) {
```

Comment:
Given the number of activity-related bug reports we get, I think this would be a good place for a console.log that indicates the decision we made.  Mainly whether we were document.hidden and whether we tried to foreground ourselves.  Especially since it's conceivable that there could be future platform regressions about us foregrounding ourselves without a user-initiated event handler/mozSetMessageHandler call directly on the stack.  (I think various guards may have been lessened and there may be some attribution that follows setTimeout calls and such, but conceivably dynamic script loading might lose the attribution and thereby cause troubles, etc. etc..